src/realm/object-store/impl/realm_coordinator.cpp

```diff
@@ -757,23 +757,15 @@ void RealmCoordinator::wait_for_change_release()
     m_db->wait_for_change_release();
 }
 
-void RealmCoordinator::pin_version(VersionID versionid)
-{
-    if (m_async_error)
-        return;
-    if (!m_advancer_sg || versionid < m_advancer_sg->get_version_of_current_transaction())
-        m_advancer_sg = m_db->start_read(versionid);
-}
-
 // Thread-safety analsys doesn't reasonably handle calling functions on different
 // instances of this type
 void RealmCoordinator::register_notifier(std::shared_ptr<CollectionNotifier> notifier) NO_THREAD_SAFETY_ANALYSIS
 {
-    auto version = notifier->version();
     auto& self = Realm::Internal::get_coordinator(*notifier->get_realm());
     {
         util::CheckedLockGuard lock(self.m_notifier_mutex);
-        self.pin_version(version);
+        if (!self.m_async_error)
+            notifier->attach_to(notifier->get_realm()->duplicate());
```

Comment:
So this is the core of the fix.

Reply:
Yes, this ended up being the only change actually needed to fix the bug. The changes in run_async_notifiers() are left over from a different attempt at fixing the bug that failed to fix it but did bring a functional benefit (of not running new notifiers with the lock held).